### ISSUE TITLE
`runtime-evaluated-decorators` not working for the annotations of decorated functions.

### ISSUE THREAD

**Issue by avandierast** [astral-sh/ruff#9312]
Hello :)

First, thank you for the great work on ruff !

I'm trying to enforce TCH003 with `quote-annotations = true` and I need to exclude some decorated functions.
But it seems that `runtime-evaluated-decorators` does not work with functions but only with classes. Or I'm doing something wrong...

A minimal reproducing code snippet:
```python
# filename: TCH003/__init__.py

from pathlib import Path
from uuid import UUID

def runtime_evaluated(obj):
    return obj

@runtime_evaluated
class Toto:
    id: UUID

@runtime_evaluated
def toto(
    id: UUID,
    path: Path,
) -> None:
    pass
```

The corresponding ruff config:
```toml
select = ["TCH003"]

[lint.flake8-type-checking]
quote-annotations = true
runtime-evaluated-decorators = ["TCH003.__init__.runtime_evaluated"]
```

Now, ruff outputs:
```bash
$ ruff --version
ruff 0.1.9
$ ruff --config .ruff.toml 
TCH003/__init__.py:3:21: TCH003 Move standard library import `pathlib.Path` into a type-checking block
Found 1 error.
No fixes available (1 hidden fix can be enabled with the `--unsafe-fixes` option).
```

The setting `runtime-evaluated-decorators` is working for classes since only `pathlib.Path` is subject to TCH003. But it does not work for the function.

The end goal of this is to make TCH00X rules work with FastAPI routers .
I found this issue but I failed to make it work : https://github.com/astral-sh/ruff/issues/6137.

Thank you for your help :)

**Comment by Divaltor** [comment]
I've messed up with same issue around `@validate_call` from `pydantic` module

It would be nice if Ruff would take such decorators into account while discovering which types should be quoted

```python
from app import User

from pydantic import validate_call

@validate_call(config={'arbitrary_types_allowed': True})
def test(user: User):
    ...
```

Converts to

```python
from typing import TYPE_CHECKING

from pydantic import validate_call

if TYPE_CHECKING:
    from app import User

@validate_call(config={'arbitrary_types_allowed': True})
def test(user: 'User'):
     ...
```

Which cause a `NameError` from `pydantic`

**Comment by charliermarsh** [comment]
It looks like we don't apply these to functions at all right now (but we probably should?).

**Comment by zanieb** [comment]
Hm the documentation for `runtime-evaluated-decorators` says

> Exempt **classes** decorated with any of the enumerated decorators from needing to be moved into type-checking blocks.

We can definitely consider expanding the scope of the setting, I do not know why it was originally limited.

It's only used at https://github.com/astral-sh/ruff/blob/2895e7d1266c661b36bbeeea31c293bdf5fdf0e7/crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs#L74-L92 right now

**Comment by charliermarsh** [comment]
PR is up :)

**Comment by DetachHead** [comment]
this seems to still be broken in ruff 0.2.2. i've raised a new issue #10089
